### PR TITLE
[JIT] Make named parameters (recursively) iterable

### LINKED ISSUE #36211: [JIT] support self.named_buffers and self.named_parameters in TorchScript
## 🚀 Feature
 support self.named_buffers and self.named_parameters in TorchScript

## Motivation

Since `self.named_models` had been supported in torchscript [#29495 ]( https://github.com/pytorch/pytorch/pull/29495), it will be more convenient to support self.named_buffers and self.named_parameters. Projects like object detection and model compression use buffers to stores important things (e.g. anchor box and pruning mask).

## Pitch
```
  class MyMod(torch.nn.Module):
      def __init__(self):
          super(MyMod, self).__init__()
          self.register_buffer('x', torch.zeros(1))

      def forward(self, x):
          for name, buffer in self.named_buffers()
              x = x + buffer
          return x

  model = MyMod()
  torch.jit.script(model)
```

### PR DESCRIPTION
**Pull request by voznesenskym**
closes #36211

### CODE REVIEW

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 1]
```diff
@@ -306,6 +306,34 @@ std::shared_ptr<SugaredDict> ModuleValue::getSugaredNamedBufferDict(
       std::make_shared<SugaredTupleValue>(values));
 }
 
+std::shared_ptr<SugaredDict> ModuleValue::getSugaredNamedParametersDict(
```
Comment (zdevito) [inline torch/csrc/jit/python/python_sugared_value.cpp:309]: I am pretty sure named_parameters, named_buffers, etc. are recursive. I don't see the code to handle that here.

Comment (voznesenskym) [reply torch/csrc/jit/python/python_sugared_value.cpp:309]: Hey, I opened #38863 to address this in a subsequent diff.

Comment (voznesenskym) [reply torch/csrc/jit/python/python_sugared_value.cpp:309]: Are you fine with me landing it as-such, and then adding the recursive implementation after?

File: torch/csrc/jit/python/python_sugared_value.h [hunk 2]
```diff
@@ -286,5 +300,121 @@ struct VISIBILITY_HIDDEN PythonClassValue : public ClassValue {
...
+      if (recurseInputValue->type()->kind() == TypeKind::BoolType) {
+        bool shouldRecurse = constant_as<bool>(recurseInputValue).value();
+        if (shouldRecurse) {
+          std::vector<SugaredValuePtr> moduleKeys;
+          std::vector<SugaredValuePtr> moduleValues;
+
+          std::vector<SugaredValuePtr> iterValues;
+          std::vector<SugaredValuePtr> iterKeys;
+
+          auto lambda = [&](std::shared_ptr<ModuleValue> m) -> void {
```
Comment (suo) [inline torch/csrc/jit/python/python_sugared_value.h:351]: `moduleKeys` and `iterKeys` seem sort of redundant here. They both have a very similar purpose: containers that are populated as we descend through the module hierarchy. I think it would be a good cleanup to refactor `recurseThroughNestedModules()` to remove the `keys` and `values` arguments. That way it implements just the traversal, and defers to the visitor lambda to actually set up whatever structures the caller expects.

**Review by suo — CHANGES_REQUESTED** [review]
Sorry for the late review! I think it's overall the right direction, I suggest some small cleanups inline.

**Comment by ppwwyyxx** [comment]
fyi I'm just importing to test the diff internally

**Comment by voznesenskym** [comment]
It’s not particularly finished ;)

File: torch/csrc/jit/python/python_sugared_value.h [hunk 3]
```diff
@@ -286,5 +300,121 @@ struct VISIBILITY_HIDDEN PythonClassValue : public ClassValue {
...
+    if (inputMap.find("prefix") != inputMap.end()) {
+      throw ErrorReport(loc)
+          << "Prefix is not currently supported yet for: " << name_ << "()";
+    }
+
+    if (inputMap.find("recurse") != inputMap.end()) {
+      auto recurseInputValue =
+          inputMap.find("recurse")->second.value(*f.graph());
+      if (recurseInputValue->type()->kind() == TypeKind::BoolType) {
+        bool shouldRecurse = constant_as<bool>(recurseInputValue).value();
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/csrc/jit/python/python_sugared_value.h:343]: Great call, didn't think of this case at all

File: torch/csrc/jit/python/python_sugared_value.h (hunk 2 above)
Comment (voznesenskym) [reply torch/csrc/jit/python/python_sugared_value.h:351]: Yeah super interesting. Let me play with what that would look like.

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 4]
```diff
@@ -245,16 +245,19 @@ void checkInterface(
...
     Function& m,
-    std::vector<SugaredValuePtr>& keys,
-    std::vector<SugaredValuePtr>& values,
     std::shared_ptr<ModuleValue> self,
     const std::string& prefix,
-    const std::string& field) {
+    const std::string& field,
+    std::function<void(
+        std::shared_ptr<ModuleValue> module,
+        SugaredValuePtr value,
```
Comment (suo) [inline torch/csrc/jit/python/python_sugared_value.cpp:253]: let's put `key` before `value` since that is more conventional

File: torch/csrc/jit/python/python_sugared_value.h [hunk 5]
```diff
@@ -286,5 +300,143 @@ struct VISIBILITY_HIDDEN PythonClassValue : public ClassValue {
...
+          inputMap.find("recurse")->second.value(*f.graph());
+      if (recurseInputValue->type()->kind() == TypeKind::BoolType) {
+        auto shouldRecurseAsConstBool = constant_as<bool>(recurseInputValue);
+        if (!shouldRecurseAsConstBool.has_value()) {
+          throw ErrorReport(loc)
+              << "Recurse boolean must be constant, not dynamically evaluted.";
+        }
+        bool shouldRecurse = shouldRecurseAsConstBool.value();
+        if (shouldRecurse) {
+          // std::vector<SugaredValuePtr> moduleKeys;
```
Comment (voznesenskym) [inline torch/csrc/jit/python/python_sugared_value.h:362]: self nit: rm
